### PR TITLE
refactor: deglobalization of bls_legacy_scheme 2/N

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
Many usages of `CBLS{Signature,PrivateKey,PublicKey}` assume using global variable, even if can be specified explicitly.
Some of these usages have been deglobalized in this PR.

Some prior improvements and fixes are here: [#5403](https://github.com/dashpay/dash/pull/5403)

## What was done?
 - Refactored the uses of global variable of `bls_legacy_scheme` from `SetHex`, `SetByteVector`, some rpc calls. 
 - Removed flag `checkMalleable` to simplify code because it's always `true`. 
 - Removed dependency from `txmempool.h` on `bls.h` to speed up compilation.

## How Has This Been Tested?
Run unit/functional tests.

## Breaking Changes
No breaking changes assumed. But in theory behaviour of some RPC can be more explicit and predictable.

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
Reindexed on testnet with no issues 👍 

Pls see some suggestions in rpc code.

**Comment by UdjinM6** [comment]
Looks good but `Check Merge Fast-Forward Only` complains... pls rebase on the recent `develop`

**Review by ogabrielides — DISMISSED** [review]
LGTM. Let's wait for CI

**Comment by UdjinM6** [comment]
hmm... so one of my suggestions was wrong https://gitlab.com/dashpay/dash/-/jobs/4505250574#L2394 and we need to partially revert 668d00db617097f1b0c63978400f9a0b6838b7ef now (pls see 140c1cb49f)

**Comment by knst** [comment]
> EDIT: On the bright side, I reindex both on mainnet and testnet with no issues :)

So far as I understand, the RPC `verifychainlock` can success (or partially success) even if we don't have this block in chain yet. So, behavior in this case indeed changed (and CI failed), but it is not a regular situation, that can easy be reproduced without intention. But it seems as `rpc_verifychainlock.py` tests it.

**Review by UdjinM6 — APPROVED** [review]
👍 

lightly tested ACK (reindexed with no issues)

File: src/evo/providertx.h [hunk 1]
```diff
@@ -167,7 +167,7 @@ class CProUpServTx
         }
         if (!(s.GetType() & SER_GETHASH)) {
             READWRITE(
-                    CBLSSignatureVersionWrapper(const_cast<CBLSSignature&>(obj.sig), (obj.nVersion == LEGACY_BLS_VERSION), true)
+                    CBLSSignatureVersionWrapper(const_cast<CBLSSignature&>(obj.sig), (obj.nVersion == LEGACY_BLS_VERSION))
```
Comment (PastaPastaPasta) [inline src/evo/providertx.h:174]: I'd request dropping b2d41f6462508b9bdbd649555c28090d4bcdc3b9 (maybe modulo this line's change) as it's not related to the deglobalization and it may be something we SHOULD use when reindexing... I'll have to check perf at some point.

File: src/evo/providertx.h [hunk 2]
```diff
@@ -149,6 +149,10 @@ class CProUpServTx
             // unknown version, bail out early
             return;
         }
+        if (obj.nVersion != BASIC_BLS_VERSION && obj.nType == MnType::HighPerformance) {
+            // for HPMN should be only BASIC bls, bail out early
+            return;
+        }
```
Comment (PastaPastaPasta) [inline src/evo/providertx.h:155]: 5989de876f34adcc672a9dfe7f8fcc270b507fb9 should probably be it's own PR... maybe even a backport?

File: src/rpc/quorums.cpp [hunk 3]
```diff
@@ -874,20 +875,37 @@ static UniValue verifychainlock(const JSONRPCRequest& request)
...
+        if (nBlockHeight < 0) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Block height out of range");
+        }
+        if (nBlockHeight <= ::ChainActive().Height()) {
+            pIndex = ::ChainActive()[nBlockHeight];
+        }
+    }
+
+    CBLSSignature sig;
+    if (pIndex) {
```
Comment (PastaPastaPasta) [inline src/rpc/quorums.cpp:898]: if we reach here, pIndex will always be non-null due to L881 or L893 (only avoided by throwing)

File: src/rpc/quorums.cpp [hunk 4]
```diff
@@ -874,20 +875,37 @@ static UniValue verifychainlock(const JSONRPCRequest& request)
...
+        bool use_legacy_signature = !llmq::utils::IsV19Active(pIndex);
+        if (!sig.SetHexStr(request.params[1].get_str(), use_legacy_signature)) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "invalid signature format");
+        }
+    } else {
+        if (!sig.SetHexStr(request.params[1].get_str(), false) &&
+                !sig.SetHexStr(request.params[1].get_str(), true)
+        ) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "invalid signature format");
+        }
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/rpc/quorums.cpp:908]: why wasn't this adopted?

File: src/llmq/dkgsession.cpp [hunk 5]
```diff
@@ -989,14 +989,15 @@ void CDKGSession::SendCommitment(CDKGPendingMessages& pendingMessages)
     qc.sig = WITH_LOCK(activeMasternodeInfoCs, return activeMasternodeInfo.blsKeyOperator->Sign(commitmentHash));
     qc.quorumSig = skShare.Sign(commitmentHash);
 
+    const bool is_bls_legacy = bls::bls_legacy_scheme.load();
```
Comment (PastaPastaPasta) [inline src/llmq/dkgsession.cpp:992]: please do this at the call site to avoid an atomic read when we basically never lie

File: src/governance/object.cpp [hunk 6]
```diff
@@ -308,7 +308,9 @@ bool CGovernanceObject::Sign(const CBLSSecretKey& key)
     if (!sig.IsValid()) {
         return false;
     }
-    vchSig = sig.ToByteVector();
+    const auto pindex = llmq::utils::V19ActivationIndex(::ChainActive().Tip());
+    bool is_bls_legacy_scheme = pindex == nullptr || nTime < pindex->pprev->nTime;
```
Comment (PastaPastaPasta) [inline src/governance/object.cpp:312]: this would need cs_main locked?

File: src/llmq/dkgsession.h [hunk 7]
```diff
@@ -7,6 +7,7 @@
 
 #include <batchedlogger.h>
 
+#include <bls/bls.h>
```
Comment (PastaPastaPasta) [inline src/llmq/dkgsession.h:10]: why?

File: src/rpc/quorums.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (knst) [reply src/rpc/quorums.cpp:908]: because it is not equal code; this suggestion changes behaviour and cause for functional tests to fail. 


Check this commit:
```
    partial revert "refactor: simplification of rpc/quorums.cpp bcs pindex can't be nullptr"
```

File: src/llmq/dkgsession.h (hunk 7 above)
Comment (knst) [reply src/llmq/dkgsession.h:10]: Because CDGKContribution has members of type `CBLSSignature`:
```
    CBLSSignature quorumSig; // threshold sig share of quorumHash+validMembers+pubKeyHash+vvecHash
    CBLSSignature sig; // single member sig of quorumHash+validMembers+pubKeyHash+vvecHash
```

This object is declared in `bls/bls.h` that is missing to be in list of includes.
Code is compilable because `bls/bls_ies.h` includes recursively `bls/bls.h`, but better to include it explicitly.

File: src/evo/providertx.h (hunk 2 above)
Comment (knst) [reply src/evo/providertx.h:155]: why do backporting only this commit but not other commits in this PR? What's a difference?

File: src/governance/object.cpp (hunk 6 above)
Comment (UdjinM6) [reply src/governance/object.cpp:312]: `::ChainActive()` locks it inside

**Comment by UdjinM6** [comment]
@knst In [ac5af56](https://github.com/dashpay/dash/pull/5443/commits/ac5af56c85643016426e401588ee098b9b360bb2) you did the opposite of what @PastaPastaPasta suggested I think. The idea was to move atomic read inside `if`/`else if` to avoid doing that outside of tests.

**Review by UdjinM6 — APPROVED** [review]
LGTM, utACK

**Comment by PastaPastaPasta** [comment]
Still not super happy here... This PR is just doing too many things imo; options are to split into a few different PRs that only really do 1 thing or get the git log to be clean enough so that it'd make sense to merge via a merge commit. I think preference is multiple PRs

**Review by ogabrielides — DISMISSED** [review]
re-ACK

**Comment by knst** [comment]
it split to https://github.com/dashpay/dash/pull/5462 and https://github.com/dashpay/dash/pull/5463

Remaining changes are rebased.

**Review by ogabrielides — DISMISSED** [review]
re-ACK

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
it looks like you (unintentionally?) dropped [some related changes](https://github.com/dashpay/dash/compare/8e6a9147ce764003c71d6ebbb827a24e7af6ed7b..c5b55d3558a78de3b12d1b7549697f36208b692a#diff-402e06df6ed3f65a5b1ea0e8a212ca593e823fcceba2b239646b1707cdde66a5R964-R984) on rebase

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by ogabrielides — APPROVED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge
